File: DevExpress VCL/ExpressQuantumGrid 6/Demos/CBuilder/ViewTableSimpleDemo/ViewTableSimpleDemoMain.cpp
//---------------------------------------------------------------------------

#include <vcl.h>
#include <shellapi.hpp>
#pragma hdrstop

#include "ViewTableSimpleDemoMain.h"
#include "ViewTableSimpleDemoData.h"
#include "dialogs.hpp"
#include "AboutDemoForm.h"
//---------------------------------------------------------------------------
#pragma package(smart_init)
#pragma link "cxClasses"
#pragma link "cxControls"
#pragma link "cxCustomData"
#pragma link "cxData"
#pragma link "cxDBData"
#pragma link "cxEdit"
#pragma link "cxFilter"
#pragma link "cxGraphics"
#pragma link "cxGrid"
#pragma link "cxGridCustomTableView"
#pragma link "cxGridCustomView"
#pragma link "cxGridDBTableView"
#pragma link "cxGridLevel"
#pragma link "cxGridTableView"
#pragma link "cxStyles"
#pragma link "cxLookAndFeels"
#pragma resource "*.dfm"
TViewTableSimpleDemoMainForm *ViewTableSimpleDemoMainForm;
//---------------------------------------------------------------------------
__fastcall TViewTableSimpleDemoMainForm::TViewTableSimpleDemoMainForm(TComponent* Owner)
  : TForm(Owner)
{
}
//---------------------------------------------------------------------------

void __fastcall TViewTableSimpleDemoMainForm::FocusRomanceCategory()
{
  for(int i=0; i < cxgGenrasDBTableView->ViewInfo->RecordsViewInfo->Count; i++) {
    if(VarToStr(cxgGenrasDBTableView->ViewData->Records[i]->Values[0]) == "Romance") {
      cxgGenrasDBTableView->ViewData->Records[i]->Focused = true;
      break;
    }
  }
}

void __fastcall TViewTableSimpleDemoMainForm::FormShow(TObject *Sender)
{
  ((TcxGridTableView*)cxgFilms->FocusedView)->DataController->Groups->FullExpand();
  FocusRomanceCategory();
}
//---------------------------------------------------------------------------

void __fastcall TViewTableSimpleDemoMainForm::miAboutClick(TObject *Sender)
{
  ShowAboutDemoForm();
}
//---------------------------------------------------------------------------

void __fastcall TViewTableSimpleDemoMainForm::miExitClick(
      TObject *Sender)
{
  Close();
}
//---------------------------------------------------------------------------

void __fastcall TViewTableSimpleDemoMainForm::miShowIndicatorClick(
      TObject *Sender)
{
  cxgFilmsDBTableView->OptionsView->Indicator = !cxgFilmsDBTableView->OptionsView->Indicator;
  ((TMenuItem*)Sender)->Checked = cxgFilmsDBTableView->OptionsView->Indicator;
}
//---------------------------------------------------------------------------

void __fastcall TViewTableSimpleDemoMainForm::miShowPreviewRowClick(
      TObject *Sender)
{
  cxgFilmsDBTableViewTAGLINE->IsPreview = !cxgFilmsDBTableViewTAGLINE->IsPreview;
  ((TMenuItem*)Sender)->Checked = cxgFilmsDBTableViewTAGLINE->IsPreview;
}
//---------------------------------------------------------------------------

void __fastcall TViewTableSimpleDemoMainForm::miFocusCellOnTabClick(
      TObject *Sender)
{
  cxgFilmsDBTableView->OptionsBehavior->FocusCellOnTab = !cxgFilmsDBTableView->OptionsBehavior->FocusCellOnTab;
  ((TMenuItem*)Sender)->Checked = cxgFilmsDBTableView->OptionsBehavior->FocusCellOnTab;
}
//---------------------------------------------------------------------------

void __fastcall TViewTableSimpleDemoMainForm::miIncSearchClick(
      TObject *Sender)
{
  cxgFilmsDBTableView->OptionsBehavior->IncSearch = !cxgFilmsDBTableView->OptionsBehavior->IncSearch;
  ((TMenuItem*)Sender)->Checked = cxgFilmsDBTableView->OptionsBehavior->IncSearch;
}
//---------------------------------------------------------------------------

void __fastcall TViewTableSimpleDemoMainForm::miImmediateEditorClick(
      TObject *Sender)
{
  cxgFilmsDBTableView->OptionsBehavior->ImmediateEditor = !cxgFilmsDBTableView->OptionsBehavior->ImmediateEditor;
  ((TMenuItem*)Sender)->Checked = cxgFilmsDBTableView->OptionsBehavior->ImmediateEditor;
}
//---------------------------------------------------------------------------

void __fastcall TViewTableSimpleDemoMainForm::miMultiSelectClick(
      TObject *Sender)
{
  cxgFilmsDBTableView->OptionsSelection->MultiSelect = !cxgFilmsDBTableView->OptionsSelection->MultiSelect;
  ((TMenuItem*)Sender)->Checked = cxgFilmsDBTableView->OptionsSelection->MultiSelect;
}

//---------------------------------------------------------------------------

void __fastcall TViewTableSimpleDemoMainForm::miShowNavigatorClick(TObject *Sender)
{
  ((TMenuItem*)Sender)->Checked = !((TMenuItem*)Sender)->Checked;
  cxgFilmsDBTableView->OptionsView->Navigator = ((TMenuItem*)Sender)->Checked;
}

File: DevExpress VCL/ExpressQuantumGrid 7/Demos/Delphi/ViewTableSimpleDemo/ViewTableSimpleDemoMain.pas
unit ViewTableSimpleDemoMain;

interface

uses
  {$IFDEF CLR}
  System.ComponentModel,
  {$ENDIF}
  Windows, Messages, Classes, Graphics, Controls, Forms, Dialogs, StdCtrls,
  cxGridLevel, cxControls, cxGridCustomView, cxGridCustomTableView, cxGridTableView,
  cxGridDBTableView, cxGrid, ExtCtrls, Grids, DBGrids, cxStyles, Menus, ActnList,
  ImgList, cxLookAndFeels, ComCtrls, cxCustomData, cxGraphics, cxFilter,
  cxData, cxEdit, DB, cxDBData, cxClasses, cxDataStorage, cxImage,
  cxTextEdit, cxMemo, cxCheckBox, cxBlobEdit, cxHyperLinkEdit,
  cxLookAndFeelPainters, dxSkinsCore, dxSkinsDefaultPainters,
  dxSkinscxPCPainter;

type
  TViewTableSimpleDemoMainForm = class(TForm)
    mmMain: TMainMenu;
    miFile: TMenuItem;
    miExit: TMenuItem;
    miOptions: TMenuItem;
    miAbout: TMenuItem;
    miShowNavigator: TMenuItem;
    pnGenres: TPanel;
    pnGenresCaption: TPanel;
    cxgGenras: TcxGrid;
    cxgGenrasDBTableView: TcxGridDBTableView;
    cxgGenrasDBTableViewNAME: TcxGridDBColumn;
    cxgGenrasLevel: TcxGridLevel;
    Splitter: TSplitter;
    pnFilms: TPanel;
    pnFilmsCaption: TPanel;
    cxgFilms: TcxGrid;
    cxgFilmsDBTableView: TcxGridDBTableView;
    cxgFilmsDBTableViewICON: TcxGridDBColumn;
    cxgFilmsDBTableViewCAPTION: TcxGridDBColumn;
    cxgFilmsDBTableViewYEAR: TcxGridDBColumn;
    cxgFilmsDBTableViewTAGLINE: TcxGridDBColumn;
    cxgFilmsDBTableViewPLOTOUTLINE: TcxGridDBColumn;
    cxgFilmsDBTableViewRUNTIME: TcxGridDBColumn;
    cxgFilmsDBTableViewCOLOR: TcxGridDBColumn;
    cxgFilmsDBTableViewPHOTO: TcxGridDBColumn;
    cxgFilmsDBTableViewWEBSITE: TcxGridDBColumn;
    cxgFilmsLevel: TcxGridLevel;
    lbDescription: TLabel;
    miShowIndicator: TMenuItem;
    miShowPreviewRow: TMenuItem;
    miFocusCellOnTab: TMenuItem;
    miImmediateEditor: TMenuItem;
    miIncSearch: TMenuItem;
    miSeparator5: TMenuItem;
    sbMain: TStatusBar;
    miMultiSelect: TMenuItem;
    LookAndFeelController: TcxLookAndFeelController;
    procedure miAboutClick(Sender: TObject);
    procedure miShowIndicatorClick(Sender: TObject);
    procedure miShowPreviewRowClick(Sender: TObject);
    procedure miFocusCellOnTabClick(Sender: TObject);
    procedure miIncSearchClick(Sender: TObject);
    procedure miImmediateEditorClick(Sender: TObject);
    procedure miMultiSelectClick(Sender: TObject);
    procedure miExitClick(Sender: TObject);
    procedure miShowNavigatorClick(Sender: TObject);
    procedure FormShow(Sender: TObject);
  private
    procedure FocusRomanceCategory;
  end;

var
  ViewTableSimpleDemoMainForm: TViewTableSimpleDemoMainForm;

implementation

{$R *.dfm}

uses
{$IFDEF CLR}
  Variants,
{$ENDIF}
  ViewTableSimpleDemoData, AboutDemoForm;

procedure TViewTableSimpleDemoMainForm.miExitClick(Sender: TObject);
begin
  Close;
end;

procedure TViewTableSimpleDemoMainForm.FocusRomanceCategory;
var
  I: Integer;
begin
  for I:=0 to cxgGenrasDBTableView.ViewInfo.RecordsViewInfo.Count - 1 do
    if cxgGenrasDBTableView.ViewData.Records[i].Values[0] = 'Romance' then
    begin
      cxgGenrasDBTableView.ViewData.Records[i].Focused := True;
      Exit;
    end;
end;

procedure TViewTableSimpleDemoMainForm.FormShow(Sender: TObject);
begin
  cxgFilms.FocusedView.DataController.Groups.FullExpand;
  FocusRomanceCategory;
end;

procedure TViewTableSimpleDemoMainForm.miAboutClick(Sender: TObject);
begin
  ShowAboutDemoForm;
end;

procedure TViewTableSimpleDemoMainForm.miShowIndicatorClick(Sender: TObject);
begin
  with cxgFilmsDBTableView.OptionsView do
    Indicator := not Indicator;
  (Sender as TMenuItem).Checked := cxgFilmsDBTableView.OptionsView.Indicator;
end;

procedure TViewTableSimpleDemoMainForm.miShowPreviewRowClick(Sender: TObject);
begin
  with cxgFilmsDBTableViewTAGLINE do
    IsPreview := not IsPreview;
  (Sender as TMenuItem).Checked := cxgFilmsDBTableViewTAGLINE.IsPreview;
end;

procedure TViewTableSimpleDemoMainForm.miFocusCellOnTabClick(Sender: TObject);
begin
  with cxgFilmsDBTableView.OptionsBehavior do
    FocusCellOnTab := not FocusCellOnTab;
  (Sender as TMenuItem).Checked := cxgFilmsDBTableView.OptionsBehavior.FocusCellOnTab;
end;

procedure TViewTableSimpleDemoMainForm.miIncSearchClick(Sender: TObject);
begin
  with cxgFilmsDBTableView.OptionsBehavior do
    IncSearch := not IncSearch;
  (Sender as TMenuItem).Checked := cxgFilmsDBTableView.OptionsBehavior.IncSearch;
end;

procedure TViewTableSimpleDemoMainForm.miImmediateEditorClick(Sender: TObject);
begin
  with cxgFilmsDBTableView.OptionsBehavior do
    ImmediateEditor := not ImmediateEditor;
  (Sender as TMenuItem).Checked := cxgFilmsDBTableView.OptionsBehavior.ImmediateEditor;
end;

procedure TViewTableSimpleDemoMainForm.miMultiSelectClick(Sender: TObject);
begin
  with cxgFilmsDBTableView.OptionsSelection do
    MultiSelect := not MultiSelect;
  (Sender as TMenuItem).Checked := cxgFilmsDBTableView.OptionsSelection.MultiSelect;
end;

procedure TViewTableSimpleDemoMainForm.miShowNavigatorClick(Sender: TObject);
begin
  with cxgFilmsDBTableView.Navigator do
    Visible := not Visible;
  (Sender as TMenuItem).Checked := cxgFilmsDBTableView.Navigator.Visible;
end;

end.
